### PR TITLE
 Capture panics during generation, Write array length (issue#61), and other fixes

### PR DESCRIPTION
**Pull request by dpordomingo**
The only one feature added with this PR is:
- 6ed3026 Capture panics to do gracefully end of generation		

This PR fixes one known issue:
- 66039ee [Capture Array length to fix issue#61](https://github.com/src-d/go-kallax/issues/61)

Minor unrelated things:
- 3339697 Add TODO for already known issues
- ecae3cd FIX Outdated JSON integration tests

### CODE REVIEW

File: generator/types.go [hunk 1]
```diff
@@ -651,6 +657,25 @@ func (f *Field) TypeSchemaName() string {
 	return parts[len(parts)-1]
 }
 
+func (f *Field) mustFindArrayLenght() int {
+	if f.Kind != Array {
+		panic(fmt.Sprintf("%s is not an Array but a %s", f.Name, f.Type))
```
Comment (erizocosmico) [inline generator/types.go]: nope, panic is completely and absolutely forbidden, return an error

File: generator/types.go [hunk 2]
```diff
@@ -651,6 +657,25 @@ func (f *Field) TypeSchemaName() string {
 	return parts[len(parts)-1]
 }
 
+func (f *Field) mustFindArrayLenght() int {
```
Comment (erizocosmico) [inline generator/types.go]: `func arrayLen(f *Field) (int, error) {`

Not all fields are arrays, so it's not something that needs to be a method of field. Also, IMHO, `arrayLen` describes as well as `findArrayLength` what the function does, while being a lot more succint.

File: generator/types.go [hunk 3]
```diff
@@ -651,6 +657,25 @@ func (f *Field) TypeSchemaName() string {
...
+	if f.Kind != Array {
+		panic(fmt.Sprintf("%s is not an Array but a %s", f.Name, f.Type))
+	}
+
+	arrayLengthError := "kallax: It could not be possible to read the array length of %s %s\n%s"
+	found := arrayLengthRegExp.FindStringSubmatch(f.Type)
+	if len(found) < 2 {
+		panic(fmt.Sprintf(arrayLengthError, f.Name, f.Type, "No legth specified"))
+	}
```
Comment (erizocosmico) [inline generator/types.go]: you can avoid all this. It's not possible that an array does not have the form `[%d]%s`

File: tests/events_test.go [hunk 4]
```diff
@@ -29,12 +29,14 @@ type eventsCheck map[string]bool
 func assertEventsPassed(s *EventsSuite, expected eventsCheck, received eventsCheck) {
 	for expectedEvent, expectedSign := range expected {
 		receivedSign, ok := received[expectedEvent]
-		if s.True(ok, fmt.Sprintf("Event '%s' was not received", expectedEvent)) {
+		if s.True(ok, fmt.Sprintf(`Event '%s' was not received
+		//TODO: https://github.com/src-d/go-kallax/issues/56
```
Comment (erizocosmico) [inline tests/events_test.go]: space after `//`

File: tests/events_test.go [hunk 5]
```diff
@@ -75,8 +77,8 @@ func (s *EventsSuite) TestEventsUpdateError() {
 
 	doc.MustFailAfter = errors.New("kallax: after")
 	updatedRows, err := store.Update(doc)
-	s.Equal(int64(0), updatedRows)
-	s.Equal(doc.MustFailAfter, err)
+	s.Equal(int64(0), updatedRows, "//TODO: https://github.com/src-d/go-kallax/issues/56")
```
Comment (erizocosmico) [inline tests/events_test.go]: same

File: tests/events_test.go [hunk 6]
```diff
@@ -75,8 +77,8 @@ func (s *EventsSuite) TestEventsUpdateError() {
...
+	s.Equal(doc.MustFailAfter, err, "//TODO: https://github.com/src-d/go-kallax/issues/56")
```
Comment (erizocosmico) [inline tests/events_test.go]: same

File: generator/types.go [hunk 7]
```diff
@@ -635,7 +640,8 @@ func (f *Field) Value() string {
 	case Slice:
 		return fmt.Sprintf("types.Slice(%s), nil", name)
 	case Array:
-		return fmt.Sprintf("types.Array(%s), nil", f.fieldVarAddress())
+		arrayLength := f.mustFindArrayLenght()
+		return fmt.Sprintf("types.Array(%s, %d), nil", f.fieldVarAddress(), arrayLength)
```
Comment (erizocosmico) [inline generator/types.go]: just call `arrayLen(f)`. the intermediate variable is not really needed for anything else

File: generator/types.go (hunk 2 above)
Comment (erizocosmico) [inline generator/types.go]: In fact, you can avoid using regexps at all: https://play.golang.org/p/lZNoERWxeb

```go
func arrayLen(f *Field) (int, error) {
	if f.Kind != Array {
		return 0, fmt.Errorf("field %s is not an array", f.Name)
	}
	
	idx := strings.Index(f.Type, "]")
	return strconv.Atoi(f.Type[1:idx])
}
```

File: generator/types.go (hunk 7 above)
Comment (erizocosmico) [inline generator/types.go]: Nvm, now it can fail

**Comment by dpordomingo** [comment]
@erizocosmico done:
5e1782d Added a space between `//` and `TODO`
ea2e085 Panic was removed

File: generator/generator.go [hunk 8]
```diff
@@ -27,13 +27,21 @@ func (g *Generator) writeFile(pkg *Package) (err error) {
 	}
 
 	defer func() {
+		if r := recover(); r != nil {
+			fmt.Println()
+			fmt.Printf("kallax: PANIC during '%s' generation:\n%s\n\n", g.filename, r)
```
Comment (erizocosmico) [inline generator/generator.go]: add a `\n` instead of an empty `fmt.Println`

**Comment by erizocosmico** [comment]
lgtm, just squash this into a single commit before merging

**Comment by dpordomingo** [comment]
rebased to maintain four different changes separately

**Pull request merged**